File: completion_list/210726/54_Spiral_Matrix.py
class Solution:

    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        def move(current, move):
            dx, dy = move
            cx, cy = current
            cx += dx
            cy += dy
            return (cx, cy)
 
        def chg_position(p):
            return (p+1) % 4
    
        m = len(matrix)
        n = len(matrix[0])
        direction = [(0,1), (1,0), (0,-1), (-1,0)]
        dp = -1
        current = (0, -1)
        result = []

        while 0<n and 0<m:
            dp = (dp+1) % 4
            for _ in range(n):
                current = move(current, direction[dp])
                cx = current[0]
                cy = current[1]
                result.append(matrix[cx][cy])
            m -= 1

            dp = (dp+1) % 4
            for _ in range(m):
                current = move(current, direction[dp])
                cx = current[0]
                cy = current[1]
                result.append(matrix[cx][cy])
            n -= 1
            
        return result

File: completion_list/210726/54_Spiral_Matrix.swift
func zip<T>(_ sequnce: [[T]]) -> [[T]] {
    guard let count = sequnce.first?.count else { return sequnce}
    return (0..<count).map { idx in sequnce.map { $0[idx] } }
}


class Solution {
    func spiralOrder(_ matrix: [[Int]]) -> [Int] {
        var matrix = matrix
        
        var answer: [Int] = []
        
        while !matrix.isEmpty {

            answer.append(contentsOf: matrix.remove(at: 0))
            matrix = zip(matrix).reversed()
            print(matrix)
        }
        
        
        return answer
    }
    
}
